src/xclient_zmq.cpp

```diff
@@ -0,0 +1,101 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Sylvain Corlay, Martin Renou          *
+* Copyright (c) 2016, QuantStack                                           *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#include "xeus-zmq/xclient_zmq.hpp"
+#include "xclient_zmq_impl.hpp"
+
+namespace xeus
+{
+
+    xclient_zmq::xclient_zmq(xcontext& context,
+                            const xeus::xconfiguration& config)
+        : p_client_impl(std::make_unique<xclient_zmq_impl>(context.get_wrapped_context<zmq::context_t>(), config))
+    {
+    }
+
+    // Has to be in the cpp because incomplete
+    // types are used in unique_ptr in the header
+    xclient_zmq::~xclient_zmq() = default;
+
+
+    void xclient_zmq::send_shell(xmessage msg)
+    {
+        p_client_impl->send_on_shell(std::move(msg));
+    }
+
+    void xclient_zmq::send_control(xmessage msg)
+    {
+        p_client_impl->send_on_control(std::move(msg));
+    }
+
+    nl::json xclient_zmq::check_shell_answer()
```

Comment:
Why not returning an optional xmessage instead of aggregating them in a json? That would avoid unecessary copies.